components/analytics-data-connectors/org.wso2.carbon.analytics.datasource.mongo/src/main/java/org/wso2/carbon/analytics/datasource/mongo/MongoAnalyticsRecordStore.java

```diff
@@ -0,0 +1,250 @@
+        if (recordGroup.getTimeFrom() == Long.MIN_VALUE && recordGroup.getTimeTo() == Long.MAX_VALUE) {
+            iterableResult = db.getCollection(collection).find().projection(fields(include("arsId", "timestamp", "data"), excludeId()));
+            if (count != -1) {
+                iterableResult = iterableResult.limit(count);
+            }
+        } else {
+            iterableResult = db.getCollection(collection).find()
+                    .filter(and(gte("timestamp", recordGroup.getTimeFrom()), lt("timestamp", recordGroup.getTimeTo())));
+            if (count != -1) {
+                iterableResult = iterableResult.limit(count);
+            }
+        }
+        return new MongoRecordDataIterator(iterableResult.iterator(), columns, tableName, tenantId);
+    }
+
+    private AnalyticsIterator<Record> readRecordsByIds(GlobalMongoRecordGroup recordGroup) throws AnalyticsException {
+        return this.lookupRecordsByIds(recordGroup.getTenantId(), recordGroup.getTableName(), recordGroup.getIds(),
+                recordGroup.getColumns());
+    }
+
+    private AnalyticsIterator<Record> lookupRecordsByIds(int tenantId, String tableName, List<String> ids, List<String> columns) {
+        String collection = this.generateTargetCollectionName(tenantId, tableName);
+        MongoCursor<Document> cursor = db.getCollection(collection).find(in("arsId", ids)).iterator();
+        return new MongoRecordDataIterator(cursor, columns, tableName, tenantId);
+    }
+
+    @Override
+    public long getRecordCount(int tenantId, String tableName, long timeFrom, long timeTo)
+            throws AnalyticsException, AnalyticsTableNotAvailableException {
+        String collection = this.generateTargetCollectionName(tenantId, tableName);
+        return db.getCollection(collection).count(and(gte("timestamp", timeFrom), lt("timestamp", timeTo)));
+    }
+
+    @Override
+    public boolean isPaginationSupported() {
+        return false;
+    }
+
+    @Override
+    public boolean isRecordCountSupported() {
+        return true;
+    }
+
+    @Override
+    public void put(List<Record> records) throws AnalyticsException, AnalyticsTableNotAvailableException {
+        Collection<List<Record>> batches = GenericUtils.generateRecordBatches(records);
+        for (List<Record> batch : batches) {
+            addBatch(batch);
+        }
+    }
+
+    private void addBatch(List<Record> batch) throws AnalyticsException, AnalyticsTableNotAvailableException {
+        Record firstRecord = batch.get(0);
+        int tenantId = firstRecord.getTenantId();
+        String tableName = firstRecord.getTableName();
+        String collection = this.generateTargetCollectionName(tenantId, tableName);
+        try {
+            Map<String, Document> documentsToInsert = new HashMap<String, Document>();
+            for (Record record : batch) {
+                if (db.getCollection(collection).count(eq("arsId", record.getId())) >= 1) {
```

Comment:
Why don't we use "_id" as the record id? .. it is already the primary key of a document, and so it's automatically indexed also. And, here a count operation before every record insertion seems a bit expensive, why don't we try to do the batch insert straightaway, and if it fails, do an update/replaceOne for all the records of that batch.